### PR TITLE
Oracle API Key generation

### PR DESCRIPTION
**Pull request by elijaholmos**
## summary
- POST `/auth` with an `oracleId` in the body to generate an apiKey
- New API key is created using `uuid` module
- entire Firestore document is returned
- previously created api key (if any) is invalidated

## other
- currently does not validate `oracleId` in the POST request

### CODE REVIEW

File: src/repositories/auth.repository.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  constructor(
+    @inject('datasources.firestore') dataSource: FirestoreDataSource,
+  ) {
+    super(OracleApiKey, dataSource);
+  }
+
+  findLatest(oracleId: number): Promise<OracleApiKey | null> {
+    return this.findOne({
+      where: {
+        oracleId,
```
Comment (kharaone) [inline src/repositories/auth.repository.ts:20]: order is expecting string[] https://loopback.io/doc/en/lb4/apidocs.filter.filter.order.html

File: src/services/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+@injectable({scope: BindingScope.TRANSIENT})
+export class AuthService {
+  constructor(
+    @repository(AuthRepository) public apiRepository: AuthRepository,
+  ) {}
+
+  /*
+   * Adapted from https://stackoverflow.com/a/8809472/8396479
+   */
+  generateApiKey() {
```
Comment (kharaone) [inline src/services/auth.service.ts:12]: why not using uuid module here ? (it's even mentioned in your SO post referenced)
the other use could be crypto.randomUUID()

But let's not reinvent the wheel here

**Comment by kharaone** [comment]
@elijaholmos Could you elaborate on the /auth endpoint ?

It seems that it's a generate/regenerate endpoint.
What happens if I pass another oracleID than mine ?
I'm sure you're not finished here yet, but that whole generate/create sequence feels weird

File: src/repositories/auth.repository.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      },
+      order: {
+        id: 'DESC',
+        //figure out how to get the highest integer ID
+      },
```
Comment (kharaone) [inline src/repositories/auth.repository.ts]:
```suggestion
      order: [ 'id DESC'],
```

**Comment by elijaholmos** [comment]
@kharaone I elected to leverage [`uuid` v1](https://www.npmjs.com/package/uuid#uuidv1options-buffer-offset) because it is timestamp-based

File: src/controllers/auth.controller.ts [hunk 4]
```diff
@@ -29,16 +29,13 @@ export class AuthController {
...
-    // find latest api key for oracle (if exists) (null check)
-    const lastApiKey = await this.authRepository.findLatest(
-      oracleApiKey.oracleId,
-    );
-    // if exists, mark as expired
-    !!lastApiKey &&
-      (await this.authRepository.updateById(lastApiKey.id, {expired: true}));
+
+    //invalidate previous oracle api key if exists
+    await this.authService.invalidateLastApiKey(oracleApiKey.oracleId); //refactor this method
```
Comment (kharaone) [inline src/controllers/auth.controller.ts]: I like this refactor.

**Review by kharaone — COMMENTED** [review]
it's definitely better now

**Review by kharaone — APPROVED** [review]
LGTM

**Pull request merged**